docs/proposals/05-etcd-operator-tasks.md

```diff
@@ -0,0 +1,364 @@
+      lastTransitionTime: <time of transition to this state>
+      reason: <reason/message if any>
+    conditions:
+      - type: <type of condition>
+        status: <status of the condition, one of True, False, Unknown>
+        lastTransitionTime: <last time the condition transitioned from one status to another>
+        reason: <programmatic identifier indicating the reason for the condition's last transition>
+        message: <human readable message indicating details about the transition>
+```
+
+> NOTE: The above custom resource YAML serves as a template which will be used to further create a CRD and Golang APIs.
+
+### Golang API
+
+```go
+// EtcdOperatorTask represents an out-of-band operator task.
+type EtcdOperatorTask struct {
+  metav1.TypeMeta
+  metav1.ObjectMeta
+
+  // Spec is the specification of the task.
+  Spec EtcdOperatorTaskSpec
+  // Status is most recently observed status of the task.
+  Status EtcdOperatorTaskStatus
+}
+```
+
+#### Spec
+
+The authors propose that the following fields should be specified in the spec (desired state) of the `EtcdOperatorTask` custom resource.
+
+* To capture the type of `out-of-band` operator task to be performed, `taskType` field should be defined in spec. It can have values of all supported "out-of-band" operations/tasks eg. "OnDemandSnaphotTask", "QuorumLossRecoveryTask" etc.
+* To capture the configuration specific to each task, a `.spec.taskConfig` field is defined in spec as type [RawExtension](https://github.com/kubernetes/apimachinery/blob/829ed199f4e0454344a5bc5ef7859a01ef9b8e22/pkg/runtime/types.go#L49-L102) as each task may have different specifications.
+
+```go
+// EtcdOperatorTaskSpec is the spec for a EtcdOperatorTask resource.
+type EtcdOperatorTaskSpec struct {
+  
+  // Type specifies the type of out-of-band operator task to be performed. 
+  Type string `json:"taskType"`
+
+  // Config is a task specific configuration.
+  Config *runtime.RawExtension `json:"taskConfig,omitempty"`
+
+  // TTLSecondsAfterFinished is the time-to-live after which the task and 
+  // related resources will be garbage collected.
+  // +optional
+  TTLSecondsAfterFinished *int32 `json:"ttlSecondsAfterFinished,omitempty"`
+}
+```
+
+#### Status
+
+The authors propose that the following fields should be specified in the `Status` (current state) of the EtcdOperatorTask custom resource as the custom resource's `Status` will be used to monitor the progress of the task.
+
+* To capture the Status of a task, a field `taskStatus` is defined in status.
+* If operation involves many stages, so to capture the status of intermediate or any stage, `.status.lastOperation` will be useful.
+
+```go
+// TaskState represents the state of the task.
+type TaskState string
+
+const (
+  Failed TaskState = "failed"
+  Pending TaskState = "pending"
+  Rejected TaskState = "rejected"
+  Completed TaskState = "completed"
+  InProgress TaskState = "inProgress"
```

Comment:
Just a wild thought, can you have a situation where you would like to cancel an InProgress task and you would like to capture the state or its equivalent to deleting a task which will show a terminating state anyways for the resource.

Reply:
Cool, may be we keep it out-of-scope and do it only if we see a real need and not just in theory. 
You can resolve this once its put in the doc.

Reply:
I see its already there with the new commit https://github.com/gardener/etcd-druid/pull/757/commits/8a497103210dd3542175944af2b627a880e421bc.
 Hence resolving it.

Reply:
We also had a discussion around it. Take this scenario:
`on-demand full snapshot` tasks has been triggered on a cluster which has lost the quorum permanently and then operator triggered the `Recover from permanent quorum loss` task, then operator should able to `abort/suspend` the previous task which he/she triggered (in this `on-demand full snapshot` ).
To solve such scenarios, we decided to go with `pre-conditions` for each task, it those pre-conditions not met then task should be `rejected`. That's why you see `rejected` state but not `suspend` state.